www/js/diary/list.js

```diff
@@ -740,5 +771,11 @@ angular.module('emission.main.diary.list',['ui-leaflet',
           $rootScope.displayingIncident = false;
         }
       });
+
+      $rootScope.$on('$stateChangeStart', function (event, toState, toStateParams, fromState, fromParams) {
+        if (fromState.url === '/enketosurvey/:form_location/:opts' && toState.url === '/diary') {
+          readAndUpdateForDay(Timeline.data.currDay);
```

Comment:
This seems super heavyweight. why not just modify the $scope variables like I do with the  mode and purpose changes? The `$scope` variables will be invalidated on the next update, but that will fill them in with the new values anyway.

Reply:
I did try with `$scope` before I jump to this. There must be something with `$state.go` and the Back button on the nav bar. Apparently, the button label does not get updated if I use `$scope` unless I refresh the entire timeline.

Reply:
I which you could help figure out a good way to refresh the data without reloading the timeline.

Reply:
can you tell me what you tried? In particular, did you use `$scope.$apply` around the scope updates?

Reply:
Here is what I did
```
                $scope.$apply(function() {
                    $scope.data.currDayTripWrappers.forEach(function(tripgj, _index, _array) {
                        $scope.populateSurveyAnswerFromTimeline(tripgj, $scope.data.unifiedConfirmsResults.surveyAnswers);
                    });
                });
```

Reply:
This sometimes make the UI freeze and sometimes the router navigate back to `diary/list.js` but nothing happens.

Reply:
I think this is because you are not actually modifying `$scope.data.unifiedConfirmsResults` when the user fills out the survey. See what I do with `mode_confirm` and `purpose_confirm`. I just set the `tripgj.userSurveyAnswer` directly.

Reply:
If you had a look inside ` $scope.populateSurveyAnswerFromTimeline`, it try to set `tripgj.userSurveyAnswer` already.

Reply:
That's true. But it does so from `$scope.data.unifiedConfirmsResults`. And `$scope.data.unifiedConfirmsResults.surveyAnswers` is populated *only* as part of `readAndUpdateForDay` - e.g. in

https://github.com/e-mission/e-mission-phone/pull/575/files#diff-9bf2100b27e218fd87648699604df6f9R922

So if you call `$scope.populateSurveyAnswerFromTimeline` on a newly filled out survey, the survey will not be in `$scope.data.unifiedConfirmsResults.surveyAnswers` and nothing will be populated.

Reply:
I also did try with this

```js
                $scope.$apply(function() {
                    $scope.data.currDayTripWrappers.forEach(function(tripgj, _index, _array) {
                        $scope.populateSurveyAnswerFromTimeline(
                            tripgj,
                            EnketoSurvey.getAllSurveyAnswers("manual/confirm_survey", { populateLabels: true })
                        );
                    });
                });
```

but still no lucks.

Reply:
This should work; I am happy to have a short call to debug why it doesn't, just like last time.
Although it is still not a perfect solution.

In particular, you read `EnketoSurvey.getAllSurveyAnswers` for each trip, although the results are not trip-specific, and that (since it uses the unified data loader), effectively makes both local and remote calls.

Why must you use `populateSurveyAnswerFromTimeline`? Why can't you set the `$scope` variables like I do with `mode_confirm`?

Reply:
Ok, I think I have figured it out.

I forgot that `EnketoSurvey.getAllSurveyAnswers` returns a promise.

```javascript
                EnketoSurvey.getAllSurveyAnswers("manual/confirm_survey", { populateLabels: true }
                ).then(function(answers) {
                    $scope.$apply(function() {
                        $scope.data.currDayTripWrappers.forEach(function(tripgj, _index, _array) {
                            $scope.populateSurveyAnswerFromTimeline(
                                tripgj,
                                answers
                            );
                        });
                    });
                });
```

Does work.

Reply:
So how should I submit the PR ? Or perhaps can you make a commit to `shankari:rk-unsw-survey` ?

Reply:
@atton16 that is a lot better since you make a single call to `getAllSurveyAnswers` for every state change. But if people are filling out multiple surveys for a particular day (maybe at the end of the day), then you will still make multiple calls to `EnketoSurvey.getAllSurveyAnswers`.

I'd still like to know why you don't just modify the in-memory variable `tripgj.userSurveyAnswer` directly. That is the obvious UX change in a controller; the user makes a change and you reflect it immediately in a `$scope` variable while also saving remotely for persistence.

Why do you have to read the saved value from the database?

Reply:
> So how should I submit the PR ? Or perhaps can you make a commit to shankari:rk-unsw-survey ?

I already asked you "Is your git-fu strong enough to pull from this PR, making the changes and generating a new PR?" What is the answer?

Reply:
> > So how should I submit the PR ? Or perhaps can you make a commit to shankari:rk-unsw-survey ?
> 
> I already asked you "Is your git-fu strong enough to pull from this PR, making the changes and generating a new PR?" What is the answer?

So I can directly commit to `shankari:rk-unsw-survey` ?

> I'd still like to know why you don't just modify the in-memory variable `tripgj.userSurveyAnswer` directly. That is the obvious UX change in a controller; the user makes a change and you reflect it immediately in a `$scope` variable while also saving remotely for persistence.
> 
> Why do you have to read the saved value from the database?

As we directed out of the page, the only was to go back is to restore to previous state. So I cannot immediately knows which `tripgj` I wanted to get. The only thing that works is to refresh the page. If remote calls is what you are worrying then maybe we can go back to `BEMUserCache`.

Reply:
> If remote calls is what you are worrying then maybe we can go back to BEMUserCache.

You cannot go back to `BEMUserCache` because it will not work for data that has already been pushed to the server - e.g. surveys from the previous day.

Reply:
> As we directed out of the page, the only was to go back is to restore to previous state. So I cannot immediately knows which tripgj I wanted to get. 

Since you are already using `$rootScope` as part of the state navigation, you can also use it pass in the `tripgj` that is effectively part of the survey state. That is what we do in the detail view.

This also has the benefit that the survey options can be modified based on trip characteristics (e.g. speed? time of day?)

Reply:
> So I can directly commit to shankari:rk-unsw-survey ?

No, you would pull from my branch into a new branch, make your changes and generate a new PR - e.g. https://stackoverflow.com/questions/9153598/how-do-i-fetch-a-branch-on-someone-elses-fork-on-github

Reply:
@atton16 if it turns out that your git-fu is not enough to strong enough to figure out to do this, I am open to making the changes to my branch and merging. But would be good to have a patch that I can apply since the changes to use `tripgj` are more complex than this small snippet.

Reply:
> Since you are already using $rootScope as part of the state navigation, you can also use it pass in the tripgj that is effectively part of the survey state. 

One caveat with this is that you cannot call `$scope.$apply` from the survey screen since you are in a different scope. Couple of potential workarounds (neither of which are that great) involve:
1. passing in the scope variable, which a bunch of the ionic components (e.g. `$ionicModal`, `$ionicPopover`) also do anyway
1. storing the result in a different variable (e.g. `tripgj.tempSurveyAnswer` and setting `surveyAnswer` back in `list.js` within a `$scope.$apply`